orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,501 @@
+              "bytes string_data field. raw_data type cannot be string.");
+
+  std::vector<std::string> res;
+  const auto& data = tensor_proto->string_data();
+  int expected_size = 1;
+  for (int i = 0; i < tensor_proto->dims_size(); ++i) {
+    expected_size *= tensor_proto->dims(i);
+  }
+
+  ORT_ENFORCE(tensor_proto->dims_size() == 0 || data.size() == expected_size, "Data size mismatch.");
+  res.insert(res.end(), data.begin(), data.end());
+  return res;
+}
+}  // namespace
+
+Status CreateOrtValuesFromTensorProtos(
+    const std::vector<const ONNX_NAMESPACE::TensorProto*>& tensor_protos,
+    NameMLValMap& name_to_ort_value) {
+  static CPUExecutionProviderInfo info;
+  static CPUExecutionProvider cpu_provider(info);
+  static AllocatorPtr cpu_allocator = cpu_provider.GetAllocator(0, OrtMemTypeDefault);
+
+  for (const auto tensor_proto : tensor_protos) {
+    TensorShape tensor_shape{utils::GetTensorShapeFromTensorProto(*tensor_proto)};
+    const DataTypeImpl* tensor_dtype = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto->data_type())->GetElementType();
+    auto p_tensor = std::make_unique<Tensor>(tensor_dtype, tensor_shape, cpu_allocator);
+    ORT_RETURN_IF_ERROR(utils::TensorProtoToTensor(Env::Default(), nullptr, *tensor_proto, *p_tensor));
+
+    OrtValue ort_value;
+    ort_value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(), DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
+    name_to_ort_value.emplace(tensor_proto->name(), ort_value);
+  }
+
+  return Status::OK();
+}
+
+Status CheckpointUtils::OrtSaveInternal(
+    const std::string& model_uri,
+    const std::vector<std::string>& trainable_param_names,
+    const PathString& checkpoint_path) {
+  auto logger_ptr = std::make_unique<logging::Logger>(logging::LoggingManager::DefaultLogger());
+  std::shared_ptr<Model> p_model;
+  ORT_RETURN_IF_ERROR(Model::Load(model_uri, p_model, nullptr, *logger_ptr));
+  Graph& graph = p_model->MainGraph();
+
+  std::vector<const ONNX_NAMESPACE::TensorProto*> trainable_weight_values;
+  trainable_weight_values.reserve(trainable_param_names.size());
+  for (size_t i = 0; i < trainable_param_names.size(); ++i) {
+    const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
+    ORT_ENFORCE(graph.GetInitializedTensor(trainable_param_names[i], tensor_proto),
+                "Failed to find weight values: ", trainable_param_names[i]);
+    trainable_weight_values.emplace_back(tensor_proto);
+  }
+
+  std::unordered_map<std::string, OrtValue> name_to_ort_value;
+  ORT_RETURN_IF_ERROR(CreateOrtValuesFromTensorProtos(trainable_weight_values, name_to_ort_value));
+
+  auto cpu_data_transfer = std::make_unique<CPUDataTransfer>();
+  DataTransferManager data_transfer_mgr;
+  auto st = data_transfer_mgr.RegisterDataTransfer(std::move(cpu_data_transfer));
+  if (!st.IsOK()) {
+    return st;
+  }
+
+  CheckpointStates states;
+  states.module_checkpoint_states.train_session_data_transfer_mgr_ = &data_transfer_mgr;
```

Comment:
so we are going from tensorproto to ortvalue  to Param back to tensorproto? Can we add a shortcut

Reply:
done